### PR TITLE
Alert component modified as notifier

### PR DESCRIPTION
**Pull request by GowthamShanmugam**
tendrl-spec-issue: Tendrl/specifications#190

Signed-off-by: GowthamShanmugam <gshanmug@redhat.com>

### CODE REVIEW

**Comment by GowthamShanmugam** [comment]
@shtripat @anmolbabu @brainfunked @r0h4n @nthomas-redhat @rishubhjain please review

File: doc/source/installation.rst [hunk 1]
```diff
@@ -50,4 +50,4 @@ Note:
 ``useradd tendrl-user -g tendrl``
 
 2. For more detailed steps please follow: 
-https://github.com/Tendrl/documentation/wiki/Tendrl-Package-Installation-Reference
\ No newline at end of file
+https://github.com/Tendrl/documentation/wiki/Tendrl-Package-Installation-Reference
```
Comment (shtripat) [inline doc/source/installation.rst]: What is changed here?

File: etc/tendrl/notifier_logging.yaml.sample [hunk 2]
```diff
@@ -16,13 +16,13 @@ handlers:
         class: logging.handlers.TimedRotatingFileHandler
         level: INFO
         formatter: simple
-        filename: /var/log/tendrl/alerting/alerting_info.log
+        filename: /var/log/tendrl/notifier/notifier_info.log
```
Comment (shtripat) [inline etc/tendrl/notifier_logging.yaml.sample]: Do we really use this yaml config now?

File: tendrl/notifier/notification/__init__.py [hunk 3]
```diff
@@ -114,12 +107,33 @@ def __init__(self):
...
-            '/notification_settings/config'
-        )
-        if alert is not None:
-            for plugin in NotificationPlugin.plugins:
-                plugin.dispatch_notification(alert)
+    def _run(self):
+        while not self.complete.is_set():
+            try:
+                alerts = get_alerts()
+                # gevent.sleep(30)
```
Comment (shtripat) [inline tendrl/notifier/notification/__init__.py]: Is this sleep required?

File: tendrl/notifier/notification/__init__.py [hunk 4]
```diff
@@ -114,12 +107,33 @@ def __init__(self):
...
+                    if not eval(alert.delivery):
+                        for plugin in NotificationPlugin.plugins:
+                            plugin.dispatch_notification(alert)
+                        update_alert_delivery(alert)
+            except(
+                ValueError,
+                KeyError,
+                etcd.EtcdException
+            )as ex:
+                Event(
```
Comment (shtripat) [inline tendrl/notifier/notification/__init__.py]: Use `log_utils`

File: tendrl/notifier/notification/handlers/mail_handler.py [hunk 5]
```diff
@@ -254,6 +233,7 @@ def dispatch_notification(self, alert):
             smtplib.SMTPSenderRefused,
             Exception
         ) as ex:
+            raise ex
```
Comment (shtripat) [inline tendrl/notifier/notification/handlers/mail_handler.py]: log and then raise

File: doc/source/installation.rst (hunk 1 above)
Comment (GowthamShanmugam) [inline doc/source/installation.rst]: some space removed i feel

File: etc/tendrl/notifier_logging.yaml.sample (hunk 2 above)
Comment (GowthamShanmugam) [inline etc/tendrl/notifier_logging.yaml.sample]: no i will remove this

File: tendrl/notifier/notification/__init__.py (hunk 3 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/notification/__init__.py]: yes  it  required i will remove from comment, i commented that for testing, i forgot to remove that #

Comment (GowthamShanmugam) [inline tendrl/notifier/notification/__init__.py]: it is required otherwise notifier will read alerting/alerts again and again, so this will used to check alerting/alert only after 30 sec once

File: tendrl/notifier/notification/handlers/mail_handler.py (hunk 5 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/notification/handlers/mail_handler.py]: ok

File: tendrl/notifier/notification/__init__.py (hunk 4 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/notification/__init__.py]: changed please check, log_utils only can use for Messages, not for ExceptionMessages so I have replaced all messages with this function call.

**Comment by GowthamShanmugam** [comment]
@shtripat changes are done

**Comment by GowthamShanmugam** [comment]
@r0h4n please review

File: etc/tendrl/notifier_logging.yaml.sample [hunk 6]
```diff
@@ -16,13 +16,11 @@ handlers:
         class: logging.handlers.TimedRotatingFileHandler
         level: INFO
         formatter: simple
-        filename: /var/log/tendrl/notifier/notifier_info.log
```
Comment (shtripat) [inline etc/tendrl/notifier_logging.yaml.sample]: I mean to remove the conf file itself if its not used :)

File: tendrl-notifier.spec [hunk 7]
```diff
@@ -31,7 +31,7 @@ rm -rf html/.{doctrees,buildinfo}
 
 %install
 %{__python} setup.py install --single-version-externally-managed -O1 --root=$RPM_BUILD_ROOT --record=INSTALLED_FILES
-install -m  0755  --directory $RPM_BUILD_ROOT%{_var}/log/tendrl/notifier
+# install -m  0755  --directory $RPM_BUILD_ROOT%{_var}/log/tendrl/notifier
```
Comment (shtripat) [inline tendrl-notifier.spec]: Why commented?

File: tendrl-notifier.spec [hunk 8]
```diff
@@ -55,7 +55,7 @@ systemctl enable tendrl-notifier
 py.test -v tendrl/notifier/tests || :
 
 %files -f INSTALLED_FILES
-%dir %{_var}/log/tendrl/notifier
+# %dir %{_var}/log/tendrl/notifier
```
Comment (shtripat) [inline tendrl-notifier.spec]: Same as above

File: etc/tendrl/notifier_logging.yaml.sample (hunk 6 above)
Comment (GowthamShanmugam) [inline etc/tendrl/notifier_logging.yaml.sample]: ahh, this file is wrong, this is not syslog, i have to change this file as syslog conf :)

File: tendrl-notifier.spec (hunk 7 above)
Comment (GowthamShanmugam) [inline tendrl-notifier.spec]: this is not required, we are not storing any log in /var/log/tendrl/notifier, so it is not required i feel

File: tendrl-notifier.spec (hunk 8 above)
Comment (GowthamShanmugam) [inline tendrl-notifier.spec]: we are not using this, so it is not required i feel

**Comment by GowthamShanmugam** [comment]
@shtripat changes done

**Comment by anmolbabu** [comment]
Can you seperate commits as:
1. Refactoring
2. New Changes

Its very huge a PR to scan otherwise if it is atleast demarcated by commits for specific purpose I think it would be easy to review with commit's purpose in mind..

File: tendrl/notifier/notification/handlers/mail_handler.py [hunk 9]
```diff
@@ -86,29 +85,11 @@ def format_message(self, alert):
         return "Subject: [Alert] %s, %s threshold breached\n\n%s" % (
             alert.resource, alert.severity, alert.tags['message'])
 
-    def get_alert_destinations(self, alert):
```
Comment (anmolbabu) [inline tendrl/notifier/notification/handlers/mail_handler.py]: Why is this removed is there an alternative??
If yes,  point me to that

File: tendrl/notifier/notification/handlers/mail_handler.py [hunk 10]
```diff
@@ -174,6 +155,18 @@ def dispatch_notification(self, alert):
         server = None
         try:
             self.set_destinations()
+            if (
```
Comment (anmolbabu) [inline tendrl/notifier/notification/handlers/mail_handler.py]: ok so I assume the logic below is the answer for https://github.com/Tendrl/notifier/pull/100/files#r133894643

But move this logic there. So, that its easy for anybody to plug a different custom logic for this task and its very modular

File: tendrl/notifier/notification/handlers/mail_handler.py (hunk 9 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/notification/handlers/mail_handler.py]: There you are checking the global notification settings like emai_brick_utilization...etc..
as we discussed earlier all users receives all alert then why we need this..

File: tendrl/notifier/objects/alert/__init__.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
+from tendrl.commons import objects
```
Comment (anmolbabu) [inline tendrl/notifier/objects/alert/__init__.py]: Why this file and a few others in this PR showing up as a new file it should clearly reflect that the file is being refactored from a different path right?? Can you just check if you have deleted this file in the old path and commited the delete..

Is this really a new code??

I would suggest you break this commit based on the purpose of the commit like:
1. Refactoring
2. New Changes
etc..

File: tendrl/notifier/notification/handlers/mail_handler.py (hunk 9 above)
Comment (anmolbabu) [inline tendrl/notifier/notification/handlers/mail_handler.py]: But now, you'll probably have something like a global index of user to subscription kind of a thing that's what @anivargi and @brainfunked were saying right.. So where are you planning to iterate that list then??

File: tendrl/notifier/objects/definition/notifier.yaml [hunk 12]
```diff
@@ -48,6 +48,9 @@ namespace.alerting:
         source:
           help: 'The process raising the alert'
           type: String
+        delivery:
```
Comment (anmolbabu) [inline tendrl/notifier/objects/definition/notifier.yaml]: I think we discussed that you would be enhancing this to handle plugin and end-point specific manner..
Have you raised an issue to track it so that we don't miss it out..

File: tendrl/notifier/notification/handlers/mail_handler.py (hunk 10 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/notification/handlers/mail_handler.py]: are you saying like i have to move set_destination() logic to get_alert_destinations()?

File: tendrl/notifier/objects/alert/__init__.py (hunk 11 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/objects/alert/__init__.py]: ok i will check this

File: tendrl/notifier/objects/definition/notifier.yaml [hunk 13]
```diff
@@ -163,51 +172,4 @@ namespace.alerting:
...
-        critical_count:
-          help: "Count of critical alerts for the cluster"
-          type: int
-        integration_id:
-          help: "Id of the cluster"
-          type: String
-      value: /clusters/{0}/alert_counters
-      list: /clusters/{0}/alert_counters
-      help: "Cluster Alert Counter"
-    NodeAlertCounters:
```
Comment (anmolbabu) [inline tendrl/notifier/objects/definition/notifier.yaml]: I guess you would still have to update alert counters to the required locations in etcd most likely to be in cluster, node , etc... namespaces itself as the same is required at the level of listing pages with no additional api calls.
How are you planning to implement that..
Cc @anivargi @gnehapk

File: tendrl/notifier/utils/central_store_util.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
+from etcd import EtcdException
+from etcd import EtcdKeyNotFound
+from tendrl.notifier.objects.alert import Alert
+from tendrl.notifier.objects.cluster_alert import ClusterAlert
+from tendrl.notifier.objects.node_alert import NodeAlert
+
+
+def read_key(key):
```
Comment (anmolbabu) [inline tendrl/notifier/utils/central_store_util.py]: Probably this you can use the one from commons commons has a function similar to this

File: version.py [hunk 15]
```diff
@@ -1 +1 @@
-__version__ = '1.4.2'
+__version__ = '1.0.0'
```
Comment (anmolbabu) [inline version.py]: Why downgraded?? :)

File: tendrl/notifier/objects/definition/notifier.yaml (hunk 13 above)
Comment (shtripat) [inline tendrl/notifier/objects/definition/notifier.yaml]: Yes this is required as per https://github.com/Tendrl/gluster-integration/issues/356. We should have alerts counts maintained at cluster, volume and node level

File: version.py (hunk 15 above)
Comment (shtripat) [inline version.py]: I feel this is new module so new version should be fine. But I leave to maintainers for final call..

Comment (GowthamShanmugam) [inline version.py]: i will change it:)

File: tendrl/notifier/objects/definition/notifier.yaml (hunk 13 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/objects/definition/notifier.yaml]: This will be done by node-agent, not from here, i feel that is the correct approach

File: tendrl/notifier/notification/handlers/mail_handler.py [hunk 16]
```diff
@@ -58,9 +60,13 @@ def get_config_help(self):
         return config_help
 
     def set_destinations(self):
-        # TODO(anmolbabu):User configuration will not just be destination email
-        # but it would also include user level subscriptions the capability
-        # underneath needs to be enhanced for that.
+        # TODO(gowtham):Get user ids from indexes who enabled
+        # email notification
```
Comment (GowthamShanmugam) [inline tendrl/notifier/notification/handlers/mail_handler.py]: @anivargi @anmolbabu here i have to find user id or name  from indexes based on who are all enabled email notification, so what is the path for that indexes and how it looks like (structure).

File: tendrl/notifier/notification/handlers/mail_handler.py [hunk 17]
```diff
@@ -58,9 +60,13 @@ def get_config_help(self):
...
+        pass
+
+    def get_alert_destinations(self):
+        # TODO(gowtham): get user emailds only who have enabled
+        # email notfication
```
Comment (GowthamShanmugam) [inline tendrl/notifier/notification/handlers/mail_handler.py]: after found user id or names who are all enabled email notification, i will pass the those details to this function to find email ids of that particular users. for now i am getting all user emails

File: tendrl/notifier/notification/handlers/mail_handler.py [hunk 18]
```diff
@@ -16,8 +16,6 @@
 
 class EmailHandler(NotificationPlugin):
 
-    handler = "smtp"
```
Comment (anmolbabu) [inline tendrl/notifier/notification/handlers/mail_handler.py]: why is this removed any specific reason??
How would you otherwise know the list of supported notifying methods... or anything of such sort

File: tendrl/notifier/notification/__init__.py [hunk 19]
```diff
@@ -51,11 +53,11 @@ def dispatch_notification(self, msg):
         raise NotImplementedError()
 
 
-class NotificationPluginManager(object):
+class NotificationPluginManager(gevent.greenlet.Greenlet):
```
Comment (anmolbabu) [inline tendrl/notifier/notification/__init__.py]: Did this work??
I remember having faced some issue around this..
Just verify once

File: tendrl/notifier/notification/__init__.py [hunk 20]
```diff
@@ -73,26 +75,17 @@ def load_plugins(self):
...
-
     def __init__(self):
         super(NotificationPluginManager, self).__init__()
         try:
             self.load_plugins()
             notification_medium = []
+            self.complete = gevent.event.Event()
             for plugin in NotificationPlugin.plugins:
                 notification_medium.append(plugin.name)
-            NS.notification_medium = notification_medium
```
Comment (anmolbabu) [inline tendrl/notifier/notification/__init__.py]: Why are you removing this?
How would a user or even tendrl know the list of supported notification medium.
May not be a critical requirement but I feel it is good to have

File: tendrl/notifier/notification/handlers/mail_handler.py (hunk 16 above)
Comment (anmolbabu) [inline tendrl/notifier/notification/handlers/mail_handler.py]: @anivargi Please provide your inputs here

File: tendrl/notifier/notification/handlers/mail_handler.py (hunk 17 above)
Comment (anmolbabu) [inline tendrl/notifier/notification/handlers/mail_handler.py]: This function was intended to get list of email ids that are to be notified of the current alert being processed which is why alert as a parameter was passed. I think you would need it when you enable a feature where users can even subscribe only to a list of alert types that is a subset of all supported alert types.

File: tendrl/notifier/utils/central_store_util.py [hunk 21]
```diff
@@ -1,8 +1,8 @@
 from etcd import EtcdException
 from etcd import EtcdKeyNotFound
-from tendrl.notifier.objects.alert import Alert
-from tendrl.notifier.objects.cluster_alert import ClusterAlert
-from tendrl.notifier.objects.node_alert import NodeAlert
+from tendrl.commons.objects.alert import Alert
+from tendrl.commons.objects.cluster_alert import ClusterAlert
+from tendrl.commons.objects.node_alert import NodeAlert
```
Comment (anmolbabu) [inline tendrl/notifier/utils/central_store_util.py]: This needs a PR on commons right?

File: etc/create_sample_job.py [hunk 22]
```diff
@@ -0,0 +1,25 @@
+import json
```
Comment (anmolbabu) [inline etc/create_sample_job.py]: Remove this file its not required

File: etc/tendrl/alerting_logging.yaml.sample [hunk 23]
```diff
@@ -0,0 +1,38 @@
+version: 1
```
Comment (anmolbabu) [inline etc/tendrl/alerting_logging.yaml.sample]: You added a wrong file and deleted the correct file below :)

File: releasenotes/source/conf.py [hunk 24]
```diff
@@ -42,7 +42,7 @@
 master_doc = 'index'
 
 # General information about the project.
-project = u'notifier Release Notes'
+project = u'alerting Release Notes'
```
Comment (anmolbabu) [inline releasenotes/source/conf.py]: why are these reverted?

File: releasenotes/source/index.rst [hunk 25]
```diff
@@ -1,5 +1,5 @@
 ============================================
- notifier Release Notes
+ alerting Release Notes
 ============================================
```
Comment (anmolbabu) [inline releasenotes/source/index.rst]: why are these reverted? and in many placed below in this commit

File: tendrl/notifier/notification/__init__.py (hunk 19 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/notification/__init__.py]: no it is working now, now i have only NotificationPluginManager gevent in notifier module

File: tendrl/notifier/notification/__init__.py [hunk 26]
```diff
@@ -111,7 +111,7 @@ def _run(self):
         while not self.complete.is_set():
             try:
                 alerts = get_alerts()
-                # gevent.sleep(30)
```
Comment (anmolbabu) [inline tendrl/notifier/notification/__init__.py]: Why this change?

File: tendrl/notifier/notification/handlers/mail_handler.py [hunk 27]
```diff
@@ -233,7 +227,6 @@ def dispatch_notification(self, alert):
             smtplib.SMTPSenderRefused,
             Exception
         ) as ex:
-            raise ex
             Event(
```
Comment (anmolbabu) [inline tendrl/notifier/notification/handlers/mail_handler.py]: You can change this also to log utils

File: tendrl/notifier/notification/__init__.py (hunk 20 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/notification/__init__.py]: for now we are not supporting user choosing a subset of alerts only to be emailed, so it is not required now

File: tendrl/notifier/utils/central_store_util.py (hunk 21 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/utils/central_store_util.py]: yes

File: tendrl/notifier/notification/handlers/mail_handler.py (hunk 9 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/notification/handlers/mail_handler.py]: i have @anivargi gave a path like email enable notification, for now he told me to get that from user dir, later it modified to get from indexes

File: tendrl/notifier/objects/definition/notifier.yaml (hunk 12 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/objects/definition/notifier.yaml]: As per we discussed we can't track the email delivery so i am keep this as flag

File: tendrl/notifier/objects/definition/notifier.yaml (hunk 13 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/objects/definition/notifier.yaml]: i have added that

File: tendrl/notifier/utils/central_store_util.py (hunk 14 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/utils/central_store_util.py]: i am changed this to get from common

File: version.py (hunk 15 above)
Comment (GowthamShanmugam) [inline version.py]: i changed

File: etc/create_sample_job.py (hunk 22 above)
Comment (GowthamShanmugam) [inline etc/create_sample_job.py]: i removed this

File: tendrl/notifier/objects/definition/notifier.yaml (hunk 12 above)
Comment (anmolbabu) [inline tendrl/notifier/objects/definition/notifier.yaml]: yep I just meant create a bug and link it here so we know what needs to be done...
But anyway you could track it using https://github.com/Tendrl/notifier/issues/29
So, that resolves the comment :)

File: etc/tendrl/alerting_logging.yaml.sample (hunk 23 above)
Comment (GowthamShanmugam) [inline etc/tendrl/alerting_logging.yaml.sample]: i changed  thi in ne commit

File: releasenotes/source/conf.py (hunk 24 above)
Comment (GowthamShanmugam) [inline releasenotes/source/conf.py]: ah sorry:)

File: releasenotes/source/index.rst (hunk 25 above)
Comment (GowthamShanmugam) [inline releasenotes/source/index.rst]: These are old commits :)

**Comment by GowthamShanmugam** [comment]
Sending another PR instead of this

File: tendrl/notifier/objects/definition/notifier.yaml (hunk 12 above)
Comment (GowthamShanmugam) [inline tendrl/notifier/objects/definition/notifier.yaml]: ok
